### ISSUE TITLE
Post mortem alternatives

### ISSUE THREAD

**Issue by spion** [nodejs/promises#8]
One of the suggested solutions to get proper post mortem core dumps is to only run promise handlers from the microtask queue in a `try/catch` block if a catch handler has already been added to the promise synchronously, when `--abort-on-unhandled-rejection` is passed as an argument.

This breaks the promise constructor as well as existing promise code. With this implemented, existing libraries are just as likely to generate core dumps as the problem code we're trying to trace down. To prevent this, node would have to add logging of warnings for unhandled rejections when catch handlers are added asynchronously - a huge shift in the current promise ecosystem.

This issue is opened to discuss alternatives that may satisfy post mortem debugging needs.

**Comment by spion** [comment]
Proposal by petka: different kind of dump that can be consumed by Chrome debugger.

The chrome debugger supports long stack traces as well as inspection of the current state of running JS code. In theory it should be possible to make v8 dump a loadable core.

Cons: 
* requires patch to V8
  * petka: might be doable if we can convince chrome developers its useful in the browser ("Send crash report to developer?")
* requires patch to chrome dev tools
* need to consider whether GC may clean up stack trace details such as function arguments
* native modules stack trace
  * would it be an acceptable compromise not to have that for promises?
* 1 tick delayed picture of data (there could be changes in theory)
  * would this be acceptable?


Pros:
* access to entire long stack, including previous events
* cross-OS, platform independent.

**Comment by petkaantonov** [comment]
I had an idea of proprietary v8 core dumps that could be opened in Chrome debugger and debugged just as well as you can debug chrome web apps live in development. Advantages that immediately come to mind over native core dump is that the debugger is javascript aware and you get long stack traces. 

However it might not be feasible to run in production due to performance and space requirement.

**Comment by benjamingr** [comment]
Don't Chrome async stack traces significantly slow things down?

**Comment by petkaantonov** [comment]
Yes and even if you are ok without having long stack traces, such mechanism still needs to be used so that proper core dump can be made after the actual original stack is gone

**Comment by petkaantonov** [comment]
Actually short stack traces might be very feasible as you don't actually need to collect them until an error happens. And long stack traces could still be separately enabled for production where performance isn't critical.

**Comment by vkurchatkin** [comment]
> I had an idea of proprietary v8 core dumps that could be opened in Chrome debugger and debugged just as well as you can debug chrome web apps live in development.

How is it different from actual core dumps? There are already v8-aware plugins for popular debuggers. Debugging with chrome tools would be nice, but it as far as I know it requires only debugging protocol implementation, nothing more.

**Comment by petkaantonov** [comment]
Chrome debugger is more accessible and familiar to web developers, it's same software that works cross platform and you can get long stack traces  rather than just the last event.

**Comment by spion** [comment]
@vkurchatkin admittedly its a far fetched idea, but basically combining some sort of a dump file with this: 

http://www.html5rocks.com/en/tutorials/developertools/async-call-stack/

**Comment by rvagg** [comment]
Related https://github.com/nodejs/node/issues/2546, it could be worthwhile connecting with the folks involved in this work (I'm assuming it's going on now, just slowly) and we may come out the other end of that process with something everyone can love.

**Comment by repenaxa** [comment]
We are about to change the way async stacks are collected in Chrome DevTools. As of today, we capture the "debug" stacks that include live scope objects and hence slow down things / leak memory.

It adds more complexity than value, so we are thinking about migrating to the v8 api for async stacks. Users would still have complete async stack chain, but would no longer be able evaluate on the async stack frames.

This makes us more conservative wrt memory use and we can make the async stack depth configurable so that clients could manage the perf overhead on their own.

**Comment by benjamingr** [comment]
@repenaxa 

 - Would the value of `this` and `arguments` be available?
 - Would it be available at the time the error happened? 
 - Would there be a way to "snapshot" the objects at the time the exception happened? (Vs when the asynchronous unhandled rejection was detected)
